core/src/main/scala/org/apache/spark/scheduler/ResultTask.scala

```diff
@@ -66,7 +67,10 @@ private[spark] class ResultTask[T, U](
     _executorDeserializeTime = System.currentTimeMillis() - deserializeStartTime
 
     metrics = Some(context.taskMetrics)
-    func(context, rdd.iterator(partition, context))
+    val itr = rdd.iterator(partition, context)
+    val result = func(context, itr)
+    val computedFullPartition = itr.isEmpty || rdd.storageLevel != StorageLevel.NONE
```

Comment:
So I've done an update so we don't depend on this used to not work with the cache + first + count example (but the new way of keeping track of when a partition is fully processed works too).